### PR TITLE
coafileTest: check last 100 commits

### PR DESCRIPTION
**Pull request by affanv14**
Fix the shortlog regex and add test to
check the regex against the last 100 commits

Fixes https://github.com/coala/coala/issues/3587

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

### CODE REVIEW

**Comment by pratyushprakash** [comment]
You have PEP8 issues. You can fix them by running `coala` before submitting a PR.

File: tests/coafileTest.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+from coalib.output.printers.LogPrinter import LogPrinter
+from pyprint.ConsolePrinter import ConsolePrinter
+from coalib.settings.ConfigurationGathering import load_config_file
+from coalib.output.printers.LOG_LEVEL import LOG_LEVEL
+
+
+class coafileTest(unittest.TestCase):
+
+    def setUp(self):
+        self.logprinter = LogPrinter(ConsolePrinter(), LOG_LEVEL.ERROR)
```
Comment (Mixih) [inline tests/coafileTest.py]: I believe this function has been deprecated and you should use `logging` instead

Comment (affanv14) [inline tests/coafileTest.py]: The function load_config_file requires logprinter.

Comment (Mixih) [inline tests/coafileTest.py]: damn, the deprecation wasn't finished yet... sigh. ok this is fine until we finally deprecate it

Comment (affanv14) [inline tests/coafileTest.py]: @Mixih So it does not work. logging.getLogger does not have warn and err methods,it has warning and error methods instead.
I think a better solution is to check the file exists beforehand,making the need for a logger unnecessary in load_config_file.

Comment (Mixih) [inline tests/coafileTest.py]: Ok, this function still needs to be fully removed. I think this is ok until we get that

**Comment by Udayan12167** [comment]
So the test file for a coafile change is not needed.

**Comment by Udayan12167** [comment]
unack 2ca8506

File: tests/coafileTest.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import os
+from coalib.misc.Shell import run_shell_command
+from coalib.settings.ConfigurationGathering import load_config_file
+
+
+class coafileTest(unittest.TestCase):
+
+    def setUp(self):
+        self.filepath = '.coafile'
+        self.assertTrue(os.path.isfile(self.filepath),
```
Comment (Mixih) [inline tests/coafileTest.py]: I don't see the point of this, plus you could have just used os.exists. Passing a logPrinter is fine for now until we get it fully removed.

**Comment by Udayan12167** [comment]
ack 3fe8824

**Comment by Mixih** [comment]
Also in this case, I think you should squash because the test file change was needed by the commit regex change so it is one change unit.

File: tests/coafileTest.py (hunk 2 above)
Comment (affanv14) [inline tests/coafileTest.py]: @Mixih the logprinter in load_config_file was only used for filenotfound exception.So i dont think it is necessary to pass a logprinter at all.It futureproofs it too.

yeah.. I think i overdid it in checking if the file exist though :)

**Comment by Udayan12167** [comment]
unack 8b5b8e1

File: tests/coafileTest.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    def setUp(self):
+        self.filepath = '.coafile'
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
+        git_command = 'git log -100 --pretty=%s'
+        git_log, stderr = run_shell_command(git_command)
+        self.assertEqual(stderr, '', msg=stderr)
+        commit_messages = git_log.split('\n')
```
Comment (Mixih) [inline tests/coafileTest.py]: splitlines?

File: tests/coafileTest.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        commit_messages = commit_messages[:100]
```
Comment (Mixih) [inline tests/coafileTest.py]: Why this slice?

Comment (affanv14) [inline tests/coafileTest.py]: damn.i had some assumptions which i just realised were wrong.okay

File: tests/coafileTest.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
+        git_command = 'git log -100 --pretty=%s'
+        git_log, stderr = run_shell_command(git_command)
+        self.assertEqual(stderr, '', msg=stderr)
+        commit_messages = git_log.splitlines()
+        for message in commit_messages:
+            self.assertIsNotNone(re.fullmatch(shortlog_regex, message),
```
Comment (Mixih) [inline tests/coafileTest.py]: why assertIsNotNone instead of assertNotEqual?

Comment (affanv14) [inline tests/coafileTest.py]: re.fullmatch returns None if there is no match.else it returns a match object.

File: tests/coafileTest.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        self.filepath = '.coafile'
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
+        git_command = 'git log -100 --pretty=%s'
+        git_log, stderr = run_shell_command(git_command)
+        self.assertEqual(stderr, '', msg=stderr)
+        commit_messages = git_log.splitlines()
+        for message in commit_messages:
```
Comment (Mixih) [inline tests/coafileTest.py]: Why don't you use git_log.splitlines() directly here? Imo, it's shorter and still clear enough

**Comment by jayvdb** [comment]
(sorry I forgot to publish the review before you uploaded a new version.)

**Comment by affanv14** [comment]
@jayvdb Do I need to change something?

**Comment by jayvdb** [comment]
I think my comments on https://github.com/coala/coala/issues/3587 and at https://github.com/coala/coala/pull/3599#discussion_r98600646 are still applicable.
The test is excellent, but I think we need more thought put into the changes to the regex.

**Comment by affanv14** [comment]
@jayvdb @Mixih how about this regex?

File: .coafile [hunk 7]
```diff
@@ -62,7 +62,7 @@ languagetool_disable_rules =
 [commit]
 bears = GitCommitBear
 shortlog_trailing_period = False
-shortlog_regex = ([^:]*|[^:]+[^ ]: [A-Z0-9*].*)
+shortlog_regex = [^:]{3,}|[^:]+[^: ]: [A-Z0-9*].{2,}
```
Comment (jayvdb) [inline .coafile]: I suggest removing this part of your PR.
The regex is still not very good.  Background reading:
https://www.google.com.au/search?q=good+commit+messages

We can merge your test now.  It is good.

Comment (affanv14) [inline .coafile]: Okay.give me a minute

**Comment by affanv14** [comment]
@jayvdb done

**Comment by sils** [comment]
this only adds the test right? The commit message tells me it'd fix the issue

**Comment by affanv14** [comment]
okay done

File: tests/coafileTest.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    def setUp(self):
+        self.filepath = '.coafile'
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
+        git_command = 'git log -100 --pretty=%s'
+        git_log, stderr = run_shell_command(git_command)
+        self.assertEqual(stderr, '', msg=stderr)
```
Comment (jayvdb) [inline tests/coafileTest.py]: assertFalse.
`msg=stderr` is unhelpful. the value of stderr will be shown anyway.

Comment (affanv14) [inline tests/coafileTest.py]: Okay.on it.

File: tests/coafileTest.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    def setUp(self):
+        self.filepath = '.coafile'
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
+        git_command = 'git log -100 --pretty=%s'
+        git_log, stderr = run_shell_command(git_command)
+        self.assertEqual(stderr, '')
```
Comment (jayvdb) [inline tests/coafileTest.py]: why didnt you use self.assertFalse as requested at https://github.com/coala/coala/pull/3599#pullrequestreview-20423278

**Comment by affanv14** [comment]
wait just a minute

**Comment by Makman2** [comment]
Actually I dislike that the test is flaky because the last 100 commits can change. With bad luck the last 100 messages are very simple, and a change in the regex gets not properly tested. Either check all commits or extract a fixed amount of them into a file/similar imo so we have always the same test conditions^^

**Comment by Makman2** [comment]
Also we have to test against bad commit messages as well^^

**Comment by affanv14** [comment]
@Makman2 How about taking the current last 350 messages and storing them in a file which is checked?

**Comment by Makman2** [comment]
@affanv14 sounds good :)

**Comment by jayvdb** [comment]
@Makman2, this isnt unit testing.  This is integration testing.
We still should have specific well designed tests for all corner cases we can think about, *especially* any bad commit messages.
This test is only to catch problems with the regex that we haven't thought about, using the existing dataset as a sanity check, because we've twice had a regex change which prevented sane commit messages, which held up PRs from being merged.

The commit messages regex should improve over time, and reject old commit messages which were not up to the new standard.  I don't mind if we store a static list of messages, but we will end up having to remove rows  occasionally.

**Comment by Makman2** [comment]
but that's better and more deterministic than not doing so. If the regex changes, our test set has to change too ;) That would also apply when we get the most recent 100 commits, if that would fail we would have to start to fetch 1 commit and increase the number when we have enough commits :)

**Comment by jayvdb** [comment]
Sure.  are you ok with "test against bad commit messages" being out of scope for this change?
If so, a stored list of 100 previous messages is the way forward (and will be quicker than fetching the last 100 out of git anyway).

**Comment by affanv14** [comment]
So take the current last 100 commits only and store them in a file?

**Comment by jayvdb** [comment]
Yup.

File: tests/coafileTest.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+class coafileTest(unittest.TestCase):
+
+    def setUp(self):
+        self.filepath = '.coafile'
+        self.commits_filepath = 'tests/commits.txt'
+        self.sections = load_config_file(self.filepath, None, True)
+
+    def test_commit_regex(self):
+        shortlog_regex = self.sections['commit']['shortlog_regex'].value
```
Comment (Makman2) [inline tests/coafileTest.py]: you could speed up the test by compiling the regex:
```python
shortlog_regex = re.compile(....value)
...
    self.assertIsNotNone(shortlog_regex.fullmatch(message))
```

File: tests/coafileTest.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        git_log = open(self.commits_filepath, 'r').read()
+        for message in git_log.splitlines():
+            self.assertIsNotNone(re.fullmatch(shortlog_regex, message),
+                                 msg='\"{0}\" does not match the '
```
Comment (Makman2) [inline tests/coafileTest.py]: you don't need to escape `"` as you use single quotes.
\+
you don't need the `0` inside `{0}` :)

File: tests/coafileTest.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+                                 'given regex'.format(message))
```
Comment (Makman2) [inline tests/coafileTest.py]: maybe also print the regex?

File: tests/coafileTest.py (hunk 11 above)
Comment (Makman2) [inline tests/coafileTest.py]: you could maybe print the `repr` version of the string, then you don't need the double quotes `"` at all and python escapes stuff probably if necessary.
You can enable this behaviour by using `{!r}`.
Up to you :)

**Comment by nemaniarjun** [comment]
@affanv14 The tests looks great! :)
But shouldn't there also be a commit which changes the shortlog?  Or is this PR just for the test?

**Comment by Makman2** [comment]
> But shouldn't there also be a commit which changes the shortlog?

What do you mean @nemaniarjun ?

**Comment by Makman2** [comment]
ah we actually need a `bad_commits.txt` file with some hand-written wrong shortlog messages. @affanv14 could you quickly craft one? :D
(btw: rename the current `commits.txt` to `good_commits.txt` :+1:)

**Comment by nemaniarjun** [comment]
@Makman2 My bad. I meant shortlog **regex**.

**Comment by Makman2** [comment]
@nemaniarjun we test whether our coafile regex works properly on shortlogs :)

**Comment by nemaniarjun** [comment]
Then we should rename the original issue. 
It misled me :stuck_out_tongue_winking_eye:  :laughing:

**Comment by Makman2** [comment]
It's just related to that issue, it does not fix it ;)

**Comment by nemaniarjun** [comment]
Aaahhh. (Makes mental to read the commit message and not the PR text)

**Comment by affanv14** [comment]
@Makman2 For bad commits,just the most common types of bad commits messages in the commit history are enough right?.Or should i add in random ones too?

**Comment by Makman2** [comment]
@affanv14 what you think is enough :)
